_drafts/tracking-code-size.md

```diff
@@ -0,0 +1,571 @@
+@click.option("--revision")
+def diff(binary, build, revision):
+    """
+    Calculate the size of the BINARY given, and compare it
+    to either the REVISION provided or the "merge-base" of the
+    current branch and master for the given BUILD.
+    """
+    text, data, bss = _calculate_codesize(binary)
+    if not revision:
+        click.echo("Using `git merge-base` against master branch")
+        # git merge-base HEAD master
+        revision = _get_merge_base("HEAD", "master")
+
+    codesize = _fetch_codesize(build, revision)
+    if not codesize:
+        click.echo("No code size for build: {}, revision: {}".format(
+            build, revision))
+        return
+
+    text_delta = text - codesize.text
+    data_delta = data - codesize.data
+    bss_delta = bss - codesize.bss
+
+    click.echo("Comparing current binary against: {}".format(revision))
+    click.echo(dict(text=text_delta, data=data_delta, bss=bss_delta))
+```
+
+Let's try it out! I went digging into the Zephyr Github Pull Requests page for a
+large changeset, and found a good candidate:
+[Timeout API rework #23382](https://github.com/zephyrproject-rtos/zephyr/pull/23382).
+
+We'll check out the branch and build the firmware first.
+
+```
+# Check out the branch for #23382
+$ cd zephyrproject/zephyr
+$ git remote add pr git@github.com:andyross/zephyr.git
+$ git fetch pr
+$ git checkout pr/timeout-take-2
+
+# Build the firmware
+$ west build -b cc3220sf_launchxl samples/net/lwm2m_client --pristine
+```
+
+Next, let's sanity check and ensure that the pull-request is based against a
+commit in the last 2,500, which would mean we had uploaded the size.
+
+```
+$ git merge-base master HEAD
+7c5db4b755f73ab5f93a6456abaa5b97bd6aff7a
+
+$ python ~/codesizes.py fetch 7c5db4b755f73ab5f93a6456abaa5b97bd6aff7a --build lwm2m_client
+{'text': 146334, 'data': 4220, 'bss': 41735}
+```
+
+Perfect! We have the commit's code size in our database. Let's get the delta.
+
+```
+$ python ~/codesizes.py diff build/zephyr/zephyr.elf --build lwm2m_client
+Using `git merge-base` against master branch
+Comparing code size of current .elf against: 7c5db4b755f73ab5f93a6456abaa5b97bd6aff7a
++---------+--------+-------+-------+
+|  Region |  .text | .data |  .bss |
++---------+--------+-------+-------+
+|   Local | 148876 |  4220 | 41735 |
+| Against | 146334 |  4220 | 41735 |
+|   Delta |   2542 |     0 |     0 |
++---------+--------+-------+-------+
+```
+
+It turns out this pull-request adds roughly 2.5kB to the master branch for our
+example firmware build! With this information in hand, the maintainers of the
+Zephyr Project can then decide whether the change is worth the bump in the
+firmware size.
+
+{:.no_toc}
+
+## Closing
+
+_All the code used in this blog post is available on
+[Github](https://github.com/memfault/interrupt/tree/master/example/faster-compilation/).
+See anything you'd like to change? Submit a pull request!_
+
+{:.no_toc}
+
+## References
+
+<!-- prettier-ignore-start -->
```

Comment:
What was prettier actually upset about here? (I usually use []() formatting  though I think both are equivalent?)

Reply:
It wraps the lines since they are longer than 80 characters and the linking breaks.

```
## References

[^zephyr]: [Zephyr Project](https://www.zephyrproject.org/)
[^zephyr_instructions]:

  [Zephyr Getting Started](https://docs.zephyrproject.org/latest/getting_started/index.html)

[^git_notes]: [Git Notes Documentation](https://git-scm.com/docs/git-notes/)
[^git_notes_post]:

  [Git Tip of the Week: Git Notes](http://alblue.bandlem.com/2011/11/git-tip-of-week-git-notes.html)

[^git_merge_base]: [git-merge-base](https://git-scm.com/docs/git-merge-base)

```

![image](https://user-images.githubusercontent.com/1041679/76882499-fc15d280-6837-11ea-8846-86bc0d83744c.png)